Collision

SteveF42 opened the pull request:
Added collision detection within the useAnimation hook.

Added a pagehash cache so only pages that change marchers are re computed saving compute time. 

Changed canvas to add little red circles depending on the collision. Currently a larger circle means marchers get close while smaller means they're basically ontop of eachother. 

note. In the future if users are allowed to change the radius of the marcher, like tenors have 10 vs snare is 5. Changing the COLLISION_RADIUS to a dynamic value should be a pretty easy change.

SteveF42 commented on apps/desktop/src/hooks/useAnimation.ts:
```diff
@@ -74,6 +88,203 @@ export const useAnimation = ({
+        calculatePageCollisions,
+    ]);
+
+    // Get collisions for the currently selected page
+    const getCollisionsForSelectedPage = useCallback(() => {
+        if (!selectedPage) {
+            return [];
+        }
+
+        const collisions = pageCollisions.get(selectedPage.id + 1) || [];
+        return collisions;
+    }, [pageCollisions, selectedPage]);
```
this is weird. selectedPage.id + 1 works here because we need to see the collisions from the current shape moving to the next. So in reality we're actually looking at the future collisions. If there's no set next in the form then there are no collisions

SteveF42 replied:
@AlexDumo oh, didn’t know that. But yeah it was my intention to get the following page.

SteveF42 commented on apps/desktop/src/hooks/useAnimation.ts:
```diff
@@ -74,6 +88,203 @@ export const useAnimation = ({
+                        const marcher2 = marcherPositionsAtTime[j];
+
+                        // find the magnitude
+                        const mag = Math.sqrt(
```
I'll try and add the Spatial Hashing algorithm this looks sick. Spatial hashing runs in O(n) so we can cut down computation by more than half lol which means we can go BLAZINGLY fast

SteveF42 replied:
also sorry for late reply I was in Mexico for the weekend

AlexDumo replied:
🇲🇽

AlexDumo reviewed (changes requested):
Some awesome stuff in here! This is almost ready to merge, just take a look at the comments I made!

As always, thanks so much for contributing :)

AlexDumo commented on apps/desktop/src/hooks/useAnimation.ts:
```diff
@@ -114,6 +121,33 @@ export const useAnimation = ({ canvas }: UseAnimationProps) => {
+        }
+
+        // this looks stupid but empty array if nothing is returned
+        const collisions =
+            pageCollisions.get(selectedPage.nextPageId || 0) || [];
```
I believe I understand? Would this not be resolved by checking if `selectedPage.nextPageId` is `null` before calling `get`?

SteveF42 replied:
yeah that's what I would think too but pageCollisions.get() can also still return undefined. It was mainly so typescript wouldn't be mad at me lol. It's probably best to change it so its more readible and add another check so undefined is never returned.


```suggestion
        const collisions = selectedPage.nextPageId
            ? pageCollisions.get(selectedPage.nextPageId)

        return collisions ?? [] // returns empty array if collisions is nullish
```

SteveF42 commented:
okay I added some dynamic time interval checking and a little bit of refactoring. I also added some information about the collisions in the inspector. 

<img width="2137" height="1386" alt="image" src="https://github.com/user-attachments/assets/85dad932-9e77-43bf-a5e3-ce87d9b49697" />

SteveF42 commented:
> Now time to review the ui,
> 
> I think the inspector panel should just be for object editing like pages and marchers. And collisions are more like show analysis/viewing/debug
> 
> maybe we could have a separate "Collisions ⚠️" tab in the toolbar when there is any collisions detected, and in there we could have it ready for future actions like "Fix collisions", but right now we could just have `5 collisions detected with marchers.` **`Review`**
> 
> Thoughts on this ux structure? I'll be ready to redesign some things if you think its ok

Yeah that sounds good. I'm assuming this would also show some dropdown or list of marchers currently colliding? You can go ahead and take it from here!
